libs/seiscomp/processing/amplitudes/iaspei.cpp

```diff
@@ -0,0 +1,168 @@
+			imax = i;
+			fmax = fabs;
+		}
+	}
+
+	return imax;
+}
+
+
+bool measure_amplitude_period(
+	std::size_t n,
+	const double *f,
+	double offset,
+	std::size_t istart,
+	std::size_t iend,
+	AmplitudePeriodMeasurement &measurement)
+{
+	// NOTE:
+	// This routine scans the input data array twice, once for the zero
+	// crossings and once for the peaks (incl. troughs!) in between.
+	// This makes the routine easier to understand but at some
+	// computational cost. Tests have shown that on a normal laptop
+	// (as of 2023) this routine is capable of making more than 100k
+	// measurements per second, so it will hardly be a bottleneck in
+	// a normal processing setup.
+
+	// Determine zero crossings.
+	std::vector<std::size_t> zero_crossings;
+	double previous { f[istart]-offset };
+	for (std::size_t i = istart+1; i < iend; i++) {
+		double current = f[i]-offset;
+		if (current*previous <= 0 && previous != 0)
+			// => zero crossing
+			// i refers to the first sample after zero crossing
+			zero_crossings.push_back(i);
+		previous = current;
+	}
+
+	if (zero_crossings.size() < 3)
+		return false;
+
+	// Determine peaks between zero crossings.
+	std::vector<std::size_t> peaks;
+	std::size_t nz = zero_crossings.size();
+	for (std::size_t k=1; k<nz; k++) {
+		std::size_t i1 = zero_crossings[k-1];
+		std::size_t i2 = zero_crossings[k];
+		std::size_t peak = find_absmax(n, f, offset, i1, i2);
+		peaks.push_back(peak);
+	}
+
+	// Keep track of the largest p2p amplitude...
+	double amax = 0;
+	// ... and the respective index
+	std::size_t kmax = 0;
+
+	std::size_t np = peaks.size();
+	for (std::size_t k = 1; k < np; k++) {
+		std::size_t i1 = peaks[k-1];
+		std::size_t i2 = peaks[k];
+
+		// New p2p maximum?
+		double a = std::abs(f[i1]-f[i2]);
+		if (a > amax) {
+			kmax = k;
+			amax = a;
+		}
+	}
+
+	// The z2p peak is the larger of the two consequtive peaks
+	// comprising the maximum p2p amplitude.
+	std::size_t ip2p1 = peaks[kmax-1];
+	std::size_t ip2p2 = peaks[kmax];
+	double ap2p1 = std::abs(f[ip2p1]-offset);
+	double ap2p2 = std::abs(f[ip2p2]-offset);
+
+	measurement.iz2p = (ap2p2 > ap2p1 ? ip2p2 : ip2p1);
+	measurement.az2p =  ap2p2 > ap2p1 ? ap2p2 : ap2p1;
+
+	measurement.ip2p1 = ip2p1;
+	measurement.ip2p2 = ip2p2;
+	measurement.ap2p1 = ap2p1;
+	measurement.ap2p2 = ap2p2;
+
+	return true;
+}
+
+
+bool measure_amplitude_period(
+	const std::vector<double> &data,
+	double offset,
+	std::size_t istart,
+	std::size_t iend,
+	AmplitudePeriodMeasurement &measurement)
+{
+	if (istart >= iend || iend >= data.size())
+		return false;
+
+	const double *f = &data[0];
+	std::size_t n { data.size() };
+
+	return measure_amplitude_period(n, f, offset, istart, iend, measurement);
+}
+
+
+double wwssnsp_amplitude_response(double frequency_hz)
+{
+	// Source:
+	// - Jim Dewey (NEIC)
+	// - IASPEI Magnitude WG.
+	// - Havskov and Ottemoeller (2010)
+	//   DOI: 10.1007/978-90-481-8697-6_6
+	//   Table 6.7
+
+	static std::vector<std::complex<double>> poles {
```

Comment:
Wouldn't it be better to expose this vector in https://github.com/SeisComP/common/blob/master/libs/seiscomp/math/filter/seismometers.cpp#L392 and reuse it here?

Reply:
Actually doing

```cpp
static WWSSN_SP_Filter wwssn_sp(Dispacement);

wwssn_sp.poles;
wwssn_sp.zeros;
wwssn_sp.norm;
```

would do the job in a less redundant way.